substrate/client/network/sync/src/strategy/state.rs

```diff
@@ -0,0 +1,758 @@
+		let unknown_hash = Hash::random();
+		let mut state_sync_provider = MockStateSync::<Block>::new();
+		state_sync_provider.expect_target_hash().return_const(target_hash);
+
+		let mut state_strategy =
+			StateStrategy::new_with_provider(Box::new(state_sync_provider), std::iter::empty());
+
+		// Unknown block imported.
+		state_strategy.on_blocks_processed(
+			1,
+			1,
+			vec![(
+				Ok(BlockImportStatus::ImportedUnknown(1, ImportedAux::default(), None)),
+				unknown_hash,
+			)],
+		);
+
+		// No actions generated.
+		assert_eq!(state_strategy.actions.len(), 0);
+	}
+
+	#[test]
+	fn succesfully_importing_target_block_finishes_strategy() {
+		let target_hash = Hash::random();
+		let mut state_sync_provider = MockStateSync::<Block>::new();
+		state_sync_provider.expect_target_hash().return_const(target_hash);
+
+		let mut state_strategy =
+			StateStrategy::new_with_provider(Box::new(state_sync_provider), std::iter::empty());
+
+		// Target block imported.
+		state_strategy.on_blocks_processed(
+			1,
+			1,
+			vec![(
+				Ok(BlockImportStatus::ImportedUnknown(1, ImportedAux::default(), None)),
+				target_hash,
+			)],
+		);
+
+		// Strategy finishes.
+		assert_eq!(state_strategy.actions.len(), 1);
+		assert!(matches!(&state_strategy.actions[0], StateStrategyAction::Finished));
+	}
+
+	#[test]
+	fn failure_to_import_target_block_finishes_strategy() {
+		let target_hash = Hash::random();
+		let mut state_sync_provider = MockStateSync::<Block>::new();
+		state_sync_provider.expect_target_hash().return_const(target_hash);
+
+		let mut state_strategy =
+			StateStrategy::new_with_provider(Box::new(state_sync_provider), std::iter::empty());
+
+		// Target block import failed.
+		state_strategy.on_blocks_processed(
+			1,
+			1,
+			vec![(
+				Err(BlockImportError::VerificationFailed(None, String::from("test-error"))),
+				target_hash,
+			)],
+		);
+
+		// Strategy finishes.
+		assert_eq!(state_strategy.actions.len(), 1);
+		assert!(matches!(&state_strategy.actions[0], StateStrategyAction::Finished));
+	}
+
+	#[test]
+	fn finished_strategy_doesnt_generate_more_actions() {
+		let target_hash = Hash::random();
+		let mut state_sync_provider = MockStateSync::<Block>::new();
+		state_sync_provider.expect_target_hash().return_const(target_hash);
+		state_sync_provider.expect_is_complete().return_const(true);
+
+		// Get enough peers for possible spurious requests.
+		let initial_peers = (1..=10).map(|best_number| (PeerId::random(), best_number));
+
+		let mut state_strategy =
+			StateStrategy::new_with_provider(Box::new(state_sync_provider), initial_peers);
+
+		state_strategy.on_blocks_processed(
+			1,
+			1,
+			vec![(
+				Ok(BlockImportStatus::ImportedUnknown(1, ImportedAux::default(), None)),
+				target_hash,
+			)],
+		);
+
+		// Strategy finishes.
+		let actions = state_strategy.actions().collect::<Vec<_>>();
+		assert_eq!(actions.len(), 1);
+		assert!(matches!(&actions[0], StateStrategyAction::Finished));
+
```

Comment:
dumb question: what would be expected result if I've called `on_block_processed` one more time here?

Reply:
`StateStrategyAction::Finished` might be generated one more time, but this won't change much. Also, it shouldn't happen in the wild as we request/import the target state only once.